### ISSUE TITLE
Upgrade to ElasticSearch 5.1

### ISSUE THREAD

**Issue by EnTeQuAk** [mozilla/addons-server#5209]
We need to uprade to ElasticSearch 5.x and are on target of 5.1 explicitly. This issue is to keep track of what's still needed to do and what's been already done about it.

* [x] Get ElasticSearch 5 to run on Travis
 * [x] Remove custom travis install script if possible
 * [x] ElasticSearch 5 by defaults has http-auth enabled, it's disabled in docker-compose but we can't change configs on travis... Make sure we use http-auth on Travis
* [x] Update docker-compose setup to run ElasticSearch 5
 * [x] Fix too much memory usage of default image
* [x] Upgrade `elasticsearch-py` and `elasticsearch-dsl` to their 5.x versions
 * [x] Fix imports, API changes
* [x] ` Deprecated field [in] used, expected [terms] instead`
* [x] `index: "no"` is not allowed anymore, use `index: False` instead
* [x] `string` datatype got replaced by `keyword` and `text`
* [x] `filtered` queries got replaced by using `bool` filters instead
* [x] `fields` got more or less replaced / deprecated / are not recommended. We should use `_source` filters instead (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-source-filtering.html)
* [x] `delete_by_query` can raise conflicts, apply `conflicts='proceed'` instead
* [x] `'type': 'phrase'` in `match` queries got deprecated, use `match_phrase` instead
* [x] Fix boosting (tags) on legacy views
* [x] Make sure all `-m es_tests` views pass!
 * [x] Fix elasticsearch related commands and tests
* [x] Fix static `_build_query` tests in `amo/test_search.py`
 * [x] Fix comments to document correct and expected filter/query usage. Might need tests verifying functionality first
* [x] `tags: {query: ['tag1', 'tag2']}` doesn't work anymore, `query` never really worked with an array before. We are boosting by tags, I added terms separately and boost them separately which should wor k for now but we'll need to watch this carefully.
* [x] Fix compat reports

Notes:

* `fuzzy query is deprecated. Instead use the [match] query with fuzziness parameter` warning won't be fixed for now, see https://github.com/elastic/elasticsearch/commit/73cf0022937ccd857da9fd0c5d5db462a982ec81 
* Travis is using ElasticSearch 5.3 for now… I am unable to install 5.1 somehow although https://github.com/elastic/elasticsearch/issues/21943#issuecomment-279815474 suggests it's possible. Needs more trial & error…

Open questions:

* Do we still care about our custom `INDEX_SETTINGS`? re https://github.com/mozilla/addons-server/issues/2661 funny enough, the elasticsearch mention that `dictionary_decompounder` is to  "decompose compound words found in many German languages" and all the words in the list are English...

**Comment by diox** [comment]
One big point: figure out a plan to handle the upgrade in production. Especially if we deploy code that is not backwards-compatible with 1.x.

**Comment by EnTeQuAk** [comment]
cc @bqbn about his plans about the upgrade in production. As far as I know we can't easily migrate to a es5 cluster and need to re-index things but I assume you may have a better overview here.

The problem here is that I'm currently not building it backwards-compatible with 1.x since that's almost impossible (since our elasticsearch libraries don't support multiple versions at the same time) so in theory, once this is done and merged we'll have to roll out es5 on -dev, -stage and -prod once the code hits these environments at the same time.

**Comment by EnTeQuAk** [comment]
From our vidyo meeting:

We do have a separate "promote" stage in our deployments, so what we'll try to do to go live is:

 * Setup a new ElasticSearch 5 Cluster
 * Push the new instances that talk to a new ElasticSearch 5 cluster but do not promote them yet
 * Run re-index (potentially without ``--with-stats`` for performance reasons) on newly created instances so that es5 cluster get's filled with data
 * Once that's done, promote new instances, test things, remove old instances
 * Maybe run re-index with ``-with-stats`` again to make sure we have up-to-date data but this should not disrupt any operations or users

Timing wise… @bqbn will be working on setting up ElasticSearch 5.x for -dev, -stage and -prod while I'm on PTO (plus minus a few days here and there depending on any issues we'll run into) and once I'm back we'll test things on all environments and potentially go live during our May 25th push.

cc @diox and @muffinresearch for any input, ideas, concerns.

**Comment by bqbn** [comment]
So on -stage, with the current 1.6 ES, the ```./manage.py reindex``` command returned quickly in about one minute. However, it left about 45K tasks in rabbitmq for celery to crunch through.

We have 9 c4.2xlarge ES nodes and 5 t2.xlarge celery nodes in -prod, whereas 3 t2.large ES nodes, and 2 t2.large celery nodes in -stage. Obviously, the two setups are not comparable.

```./manage.py reindex --with-stats``` on -stage seems to have the same issue as in -dev, i.e. the command spins forever and won't return (see https://github.com/mozilla/addons-server/issues/3960 for the decription of the issue).

I plan to try import the ad-block data from -prod to -stage and try ```reindex --with-stats``` again tomorrow or early next week.

**Comment by bqbn** [comment]
After importing some production stats data of addon id 1865 to -stage, I ran ```./manage.py reindex --with-stats``` again on -stage, and unfortunately, the command still stuck in some endless loop, and didn't want to return successfully. 

The tables that were used to dump data from -prod are:
```
'stats_addons_collections_counts'
'stats_collections_counts'
'stats_collections'
'download_counts'
'update_counts'
'theme_update_counts'
'theme_update_counts_bulk'
'stats_contributions'
'global_stats'
'theme_user_counts'
```

Are there other means we can use to verify that ```./manage.py reindex --with-stats``` will be successful on -prod? My understanding is that running this command is a must-do step for upgrading the ES cluster. 

A couple notes:
1. After waiting till all rabbitmq tasks are consumed, if I manually remove the 2 entries in ```zadmin_reindexing``` table, the ```reindex --with-stats``` command will return, however, the aliases are not pointed to the new indices.
2. It looks like our last ```reindex --with-stats``` run on -prod was also a couple years ago. 
```
$ curl http://localhost:9200/_aliases?pretty
{
  "addons_stats_prod-20151106124551" : {
    "aliases" : {
      "addons_stats_prod" : { }
    }
  },
  ... ...
```

**Comment by bqbn** [comment]
I just ran the "manage.py reindex --with-stats" again, using the latest commit in the try-elasticsearch-latest branch (04825539bc15a499770ccbff996e56f619d94e80).

The command still doesn't return even after all tasks in rabbitmq have been processed by celery workers. In the end, I had to delete the entries in zadmin_reindexing table to make the command return. 

It may be worth noting that this time there were only 3 "out of sync" errors, which, iirc, is less than previous runs. iirc. One of them looks 

One that happened at 2017-06-05T05:57:30 UTC is this:

```
Celery TASK exception: ProgrammingError: (2014, "Commands out of sync; you can't run this command now") :/data/olympia/src/olympia/amo/celery.py:61
Traceback (most recent call last):
File "/usr/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
R = retval = fun(*args, **kwargs)
File "/usr/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
return self.run(*args, **kwargs)
File "/data/olympia/src/olympia/lib/es/management/commands/reindex.py", line 77, in index_data
get_modules()[alias].reindex(index)
File "/data/olympia/src/olympia/stats/search.py", line 158, in reindex
index_stats(index_name)
File "/data/olympia/src/olympia/stats/management/commands/index_stats.py", line 88, in index_stats
'%s__range' % date_field: date_range})),
File "/usr/lib/python2.7/site-packages/django/db/models/query.py", line 162, in __iter__
self._fetch_all()
File "/usr/lib/python2.7/site-packages/django/db/models/query.py", line 965, in _fetch_all
self._result_cache = list(self.iterator())
File "/usr/lib/python2.7/site-packages/django/db/models/query.py", line 1217, in iterator
for row in compiler.results_iter():
File "/usr/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 794, in results_iter
results = self.execute_sql(MULTI)
File "/usr/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 842, in execute_sql
cursor.close()
File "/usr/lib64/python2.7/site-packages/MySQLdb/cursors.py", line 84, in close
while self.nextset():
File "/usr/lib64/python2.7/site-packages/MySQLdb/cursors.py", line 172, in nextset
nr = db.next_result()#012ProgrammingError: (2014, "Commands out of sync; you can't run this command now")
```

**Comment by EnTeQuAk** [comment]
Hmm, so unfortunately the traceback didn't change, I hoped the closed cursor fix fixed the problem or at least would lead us to another issue on the way. I'll continue investigating.

**Comment by bqbn** [comment]
I ran "reindex --with-stats" again using the latest commit (7554e1b7135af40e920efe62ec343ffa96acfbc8). The command still wasn't able to return, and it recorded the following 3 retries.
```
Jun 16 20:49:02 ip-172-31-36-125 http_app_addons_dev: [][] celery.worker.job:INFO Task olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] retry: Retry in 180s: ProgrammingError(2014, "Commands out of sync; you can't run this command now") :/usr/lib/python2.7/site-packages/celery/worker/job.py:416
Jun 16 21:26:52 ip-172-31-36-125 http_app_addons_dev: [][] celery.worker.job:INFO Task olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] retry: Retry in 180s: ProgrammingError(2014, "Commands out of sync; you can't run this command now") :/usr/lib/python2.7/site-packages/celery/worker/job.py:416
Jun 16 22:02:38 ip-172-31-3-102 http_app_addons_dev: [][] celery.worker.job:INFO Task olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] retry: Retry in 180s: ProgrammingError(2014, "Commands out of sync; you can't run this command now") :/usr/lib/python2.7/site-packages/celery/worker/job.py:416
```

A few notes about this run,
1. I noticed that for each retry, instead of continuing from where the exception happened, the program added the full set of jobs into the queue. At the beginning there was 2K+ jobs, then after the first exception (thus the first retry) happened, the queue size was increased to 4K+, then after second retry, it became 6K+, then 8K+ jobs after the third retry. You can tell the exception happened relatively quickly because for example, if the exceptions happened half way through processing the stats queue, then the numbers would have been 2K+, then 3K+, 4K+, 5K+ for each retry.

2. All the failed logs are related to this job ID: ```0de25258-ab70-427a-8bba-5ef488251d55```. I tracked down the ID to this job:
    ```
    Received task: olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] :/usr/lib/python2.7/site-packages/celery/worker/strategy.py:55
    ```
    The full logs related to this ID can be found here: https://kibana-shared-us-east-1.stage.mozaws.net/#dashboard/temp/AVy1AKoj-AxpE_GC1N7Q

3. There are 3 "soft time limit exceeded" warnings in the logs, which I believe caused the 3 exceptions (thus 3 retries). Looking at the warnings below, I'm not sure if this means the ```index_data``` function itself was also counted as a job, and thus celery workers were expecting the function to return within 1800 seconds. If that's the case, then 1800s is definitely not enough since the function needs more time to reindex all stats. 

```
Jun 16 20:44:31 ip-172-31-36-125 http_app_addons_dev: [][] celery.worker.job:WARNING Soft time limit (1800s) exceeded for olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] :/usr/lib/python2.7/site-packages/celery/worker/job.py:365
Jun 16 21:22:02 ip-172-31-36-125 http_app_addons_dev: [][] celery.worker.job:WARNING Soft time limit (1800s) exceeded for olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] :/usr/lib/python2.7/site-packages/celery/worker/job.py:365
Jun 16 22:35:38 ip-172-31-36-125 http_app_addons_dev: [][] celery.worker.job:WARNING Soft time limit (1800s) exceeded for olympia.lib.es.management.commands.reindex.index_data[0de25258-ab70-427a-8bba-5ef488251d55] :/usr/lib/python2.7/site-packages/celery/worker/job.py:365
```

**Comment by EnTeQuAk** [comment]
Thanks a lot @bqbn for running this again, I'll look into the results today and try to find more ways to reproduce the "out of sync"-errors 🤔

**Comment by bqbn** [comment]
Success! ✋ This time "reindex --with-stats" was able to return. And the two aliases got created properly as well.

```
$ curl elasticsearch-amodev1-20170622-2.amo.us-east-1.dev.mozaws.net:9200/_aliases?pretty
{
  "addons_stats_dev-20170623191340" : {
    "aliases" : {
      "addons_stats_dev" : { }
    }
  },
  "addons_dev-20170623191340" : {
    "aliases" : {
      "addons_dev" : { }
    }
  }
}

$ curl elasticsearch-amodev1-20170622-2.amo.us-east-1.dev.mozaws.net:9200/_cat/indices
green open addons_stats_dev-20170623191340 26NCtmTBRdis4z9RjAt1-g 5 2 81456439 1944379 23.4gb 7.7gb
green open addons_dev-20170623191340       iYqYB1aGSPGw-g7FXKvJNw 5 2  6068521       0  7.7gb 2.5gb
```

There were 18 warnings about clocks out of sync during the run. Here is an example:

```
Jun 23 19:21:20 ip-172-31-11-176 http_app_addons_dev: [][] celery.events.state:WARNING Substantial drift from celery@ip-172-31-33-204 may mean clocks are out of sync.  Current drift is#01217 seconds.  [orig: 2017-06-23 19:21:20.470999 recv: 2017-06-23 19:21:03.727854]#012
 :/usr/lib/python2.7/site-packages/celery/events/state.py:74
```

Here is the link to the warning logs: https://kibana-shared-us-east-1.stage.mozaws.net/#dashboard/temp/AVzX75Tt-AxpE_GCimtF

I checked the time on the two celery workers are correct. I'm not sure what it complains about, or if this is important.

**Comment by EnTeQuAk** [comment]
> Success! ✋ This time "reindex --with-stats" was able to return. And the two aliases got created properly as well.

🎉 ✋ 

> There were 18 warnings about clocks out of sync during the run. Here is an example:

Are the times on all servers, web, rabbitmq and workers identical without any drift to ntp? In any case, it's not an issue here and generally should not be for most of our tasks but should be something to check, e.g to make sure our web servers don't see any weird timing drifts so that API-auth isn't affected.

Let's talk on Tuesday when to merge and push all these changes! I'd suggest - if that works for you - let's merge after the tag and push it next week?

**Comment by bqbn** [comment]
Nope, afaict, the time on those servers is not drifted.

**Comment by EnTeQuAk** [comment]
QA:

This change covers many parts of AMO, first and foremost our search which should continue work properly and no jumpy ranking changes should be noticable. Smaller ranking changes here and there e.g for similar or identical names are expected.

Make sure that addons-frontend works as expected, ElasticSearch is entirely used throughout the APIs that are used for addons-frontend.

Let me know if you have any questions!

**Comment by bqbn** [comment]
I was checking the logs of the new elasticsearch cluster and see a lot warnings such as

```
[2017-07-26T14:11:16,878][WARN ][o.e.d.c.ParseField       ] Deprecated field [exclude] used, expected [excludes] instead
```
Not sure if this means we should change to use `excludes` everywhere `exclude` is used.

Just FYI.

**Comment by EnTeQuAk** [comment]
So, this got merged now and is on -dev and -stage. Apparently I did not add a proper "fixed" message to the commit.
